File: 1335-Minimum Difficulty of a Job Schedule.py
class Solution:
    def minDifficulty(self, jobDifficulty: List[int], d: int) -> int:
        if len(jobDifficulty) < d:
            return -1
        N = len(jobDifficulty)
        
        @cache
        def dfs(startIdx, d):
            if d == 1:
                return max(jobDifficulty[startIdx:])
            
            mx = 0
            res = float('inf')
            for i in range(startIdx, N):
                if i == N - d + 1:
                    break
                mx = max(mx, jobDifficulty[i])
                res = min(res, mx + dfs(i+1, d-1) )
                
            return res
        
        return dfs(0, d)

File: C++/1335-Minimum Difficulty of a Job Schedule.cpp
class Solution {
public:
    int N;
    int cal(vector<int>& jobDifficulty, vector<vector<int>>& memo, int idx, int remain){
        if (idx == N) return 0;
        if (remain <= 0) return INT_MAX/2;
        if (memo[idx][remain] != -1) return memo[idx][remain];
        // put jobs
        int currMax = 0;
        int res = INT_MAX/2;
        for (int i=idx; i<N; i++){
            if (N-i < remain) break;
            currMax = max(currMax, jobDifficulty[i]);
            res = min(res, currMax + cal(jobDifficulty, memo, i+1, remain-1));
        }
        return memo[idx][remain] = res;
    }
    
    int minDifficulty(vector<int>& jobDifficulty, int d) {
        this->N = jobDifficulty.size();
        vector<vector<int>> memo(N+1, vector<int>(d+1, -1));
        if (N < d)  return -1;
        
        return cal(jobDifficulty, memo, 0, d);
    }
};